Add Rainbow DQN

nuance1979 opened the pull request:
- Implement Rainbow DQN
- Reference: [paper](https://arxiv.org/abs/1710.02298) and [code1](https://github.com/Kaixhin/Rainbow), [code2](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch)

I am currently running Atari examples. Will update the results soon.

Trinkle23897 commented on docs/index.rst:
```diff
@@ -13,6 +13,7 @@ Welcome to Tianshou!
 * :class:`~tianshou.policy.DQNPolicy` `Double DQN <https://arxiv.org/pdf/1509.06461.pdf>`_
 * :class:`~tianshou.policy.DQNPolicy` `Dueling DQN <https://arxiv.org/pdf/1511.06581.pdf>`_
 * :class:`~tianshou.policy.C51Policy` `Categorical DQN <https://arxiv.org/pdf/1707.06887.pdf>`_
+* :class:`~tianshou.policy.C51Policy` `Rainbow DQN <https://arxiv.org/pdf/1707.02298.pdf>`_
```
```suggestion
* :class:`~tianshou.policy.RainbowPolicy` `Rainbow DQN <https://arxiv.org/pdf/1707.02298.pdf>`_
```

Do we need to create such a policy class? Or at least provide a `test_rainbow.py`.

nuance1979 replied:
It seems that we can get away with reusing `C51Policy` because all the other aspects of Rainbow DQN are done in either the network or the buffer. I think this is a testimony of the great design of tianshou classes, which separates concerns nicely.

I added `test_rainbow.py` for unit testing.

Trinkle23897 commented on tianshou/data/buffer/prio.py:
```diff
@@ -85,3 +85,6 @@ def __getitem__(self, index: Union[slice, int, List[int], np.ndarray]) -> Batch:
         batch = super().__getitem__(indice)
         batch.weight = self.get_weight(indice)
         return batch
+
+    def update_beta(self, beta: float) -> None:
+        self._beta = beta
```
why?

nuance1979 replied:
In the Rainbow paper Table 1, it does mention annealing beta from 0.4 to 1.0 . Many reference implementations did it. However, I did some experiments and found little difference, at least on Atari games. So I'm getting rid of it for now.

Trinkle23897 reviewed (commented):
Also could you please check https://github.com/thu-ml/tianshou/issues/393?

nuance1979 commented on examples/atari/README.md:
```diff
@@ -82,6 +82,20 @@ One epoch here is equal to 100,000 env step, 100 epochs stand for 10M.
+| MsPacmanNoFrameskip-v4      | 2419        | ![](results/rainbow/MsPacman_rew.png)     | `python3 atari_rainbow.py --task "MsPacmanNoFrameskip-v4"`  |
+| SeaquestNoFrameskip-v4      | 590         | ![](results/rainbow/Seaquest_rew.png)     | `python3 atari_rainbow.py --task "SeaquestNoFrameskip-v4"`  |
+| SpaceInvadersNoFrameskip-v4 | 645.5       | ![](results/rainbow/SpaceInvaders_rew.png) | `python3 atari_rainbow.py --task "SpaceInvadersNoFrameskip-v4"`  |
+
```
I've been trying to figure this out for the past week. Fixed a few bugs but the results are still pretty bad. A few observations:
- Unlike other methods, where the test reward climbs up, there is a lot of thrashing during the training, i.e., a high test reward followed by a very low test reward then high again;
- The training loss is so small that I'm worried that we might be encountering numerical issues. This is due to the fact that our C51 implementation computes `softmax` instead of `log_softmax`. See reference https://github.com/Kaixhin/Rainbow/blob/9ff5567ad1234ae0ed30d8471e8f13ae07119395/agent.py#L66 . However, it seems the C51 results are still quite reasonable so I'm not sure.

I attached a screenshot of the symptoms. What do you think?

<img width="1076" alt="Screen Shot 2021-07-08 at 9 02 19 AM" src="https://user-images.githubusercontent.com/10931178/124954766-4a218d00-dfcb-11eb-8de7-67a1786f3dec.png">

Trinkle23897 replied:
have you tried both prio-buffer and normal buffer?

nuance1979 replied:
You might be onto something. I tried to disable one aspect of Rainbow and got the following:

- disable dueling: 3797.5 -> 3070.0
- disable noisy net: 3797.5 -> 3850.0
- disable prio-buffer: 3797.5 -> 7830.0

So it seems that prio-buffer is making things worse. I previously did a sanity check run where I disabled all three and got back the C51 results. I'll do it again since I've changed some code since then.

Have you checked the correctness of prio-buffer before? Do you think there might be a bug or just hyper parameter misalignment? Currently I'm using `--alpha 0.5 --beta 1`. Recall we had a discussion above that the paper and the reference code mentioned annealing `beta` from 0.4 to 1 but my previous experiments didn't show any visible difference.

Trinkle23897 replied:
I'll check it tonight. Btw, have you tried to use only 1 env (and using PrioReplayBuffer instead of vectorized version)?

nuance1979 replied:
No, I haven't. I'll give it a try.

Trinkle23897 replied:
Btw if you have tried to align with sb2 prio-buffer, here's the issue: https://github.com/hill-a/stable-baselines/issues/941

Trinkle23897 replied:
I think it is related to 3 aspects:

1. policy's learn function `batch.weight` calculation, to send to `buffer.update_weight`;
2. single prio-buffer's correctness and performance;
3. vectorized prio-buffer's correctness and performance.

nuance1979 replied:
I ran several experiments over the weekend and here are the results on Qbert so far:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| current rainbow | Y | Y | Y | 3797.5 |
| no prio                 | N | Y | Y  | 7830   |
| no noisy               | Y | N | Y | 3850 |
| no dueling          | Y | Y | N | 3070 |
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| single prio           | Y | Y | Y | 4240 |
| single prio only   | Y | N | N | 4962.5 |

My conclusions:
- Priority buffer is the main culprit; from "sanity check" (=C51) to "prio only" (=C51 w/ prio): 16530 -> 3982.5;
- Single vs vectorized might have some difference but should not be the main issue.

I'll try removing the eps hack and see what happens.

nuance1979 replied:
I got some results back. See below:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/o eps hack (beta=1) | Y | N | N | 770 |
| w/o eps hack + beta=0.7 | Y | N | N | 3950 |
| w/o eps hack + beta=0.4 | Y | N | N | 4957.5 |
| single prio only   | Y | N | N | 4962.5 |
| w/o eps hack (beta=1) | Y | N | N | 692.5 |
| current rainbow | Y | Y | Y | 3797.5 |
| w/o eps hack (beta=1) | Y | Y | Y | 1000 |
| w/o eps hack + beta=0.4 | Y | Y | Y | 5090 |

I don't know why "removing eps hack + beta=1" got the worst result but "removing eps hack + beta=0.4" is currently the best. Single prio buffer's results seemed to follow vectorized prio buffer closely.

I'm running more experiments with noisy_std=0.5 (currently noisy_std=0.1 following Kaixhin's setup).

nuance1979 replied:
I have another idea: instead of getting rid of eps hack, which protects against div-by-zero exception, I can scale the weights by a constant, such as `1e9` so that adding eps (~`1e-7`) wouldn't affect the weights significantly.

Trinkle23897 replied:
But have you carefully examined the prio-buffer and align with other (rainbow) successful implementations? (Not sure what's happening but I've tried my best to check the buffer is correct)

And one more thing: have you use single env with update_freq= 4 env_step?

nuance1979 replied:
> But have you carefully examined the prio-buffer and align with other (rainbow) successful implementations? (Not sure what's happening but I've tried my best to check the buffer is correct)
> 
No, I haven't. I believe that you most likely know more about it than I do. If you can't find any major error, neither can I.

> And one more thing: have you use single env with update_freq= 4 env_step?
Do you mean `atari_rainbow.py --target-update-freq=4`? (current default 500.)

Trinkle23897 replied:
No i mean currently 10 env step once and perform update, that's 10:1 env_step:grad_step, and in traditional dqn setting it is 4:1

Trinkle23897 replied:
> If you can't find any major error, neither can I.

Don't say that, the more people check, the more reliable the code is.

nuance1979 replied:
> No i mean currently 10 env step once and perform update, that's 10:1 env_step:grad_step, and in traditional dqn setting it is 4:1

I see. So it's `atari_rainbow.py --training-num 4`, right? Sure I can try it.

nuance1979 replied:
Weight scaling seems to be working very well:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |

Trinkle23897 replied:
> I see. So it's atari_rainbow.py --training-num 4, right? Sure I can try it.

Yep

Trinkle23897 commented on test/discrete/test_rainbow.py:
```diff
@@ -0,0 +1,187 @@
+    parser.add_argument('--test-num', type=int, default=100)
+    parser.add_argument('--logdir', type=str, default='log')
+    parser.add_argument('--render', type=float, default=0.)
+    parser.add_argument('--prioritized-replay',
+                        action="store_true", default=False)
```
because the default option is False

Trinkle23897 commented on tianshou/policy/modelfree/c51.py:
```diff
@@ -95,7 +95,7 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
         loss = (cross_entropy * weight).mean()
         # ref: https://github.com/Kaixhin/Rainbow/blob/master/agent.py L94-100
-        batch.weight = cross_entropy.detach()  # prio-buffer
+        batch.weight = cross_entropy.detach() * 1e9  # prio-buffer
```
But I still don't quite get why it works. I know when using prio-buffer the current implementation will quickly reach loss≈0

nuance1979 replied:
I suspect that other implementations have done some weight scaling on the fly. And I think I might have found the evidence:

Here the weights returned are normalized by max:
https://github.com/Kaixhin/Rainbow/blob/master/memory.py#L154
```python
weights = torch.tensor(weights / weights.max(), dtype=torch.float32, device=self.device)  # Normalise by max importance-sampling weight from batch
```

I'm running an experiment where I get ride of the `1e9` scaling on L98 but add a `weight = weight / weight.max()` in `prio.py`.

nuance1979 replied:
Here are the results:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |
| w/ weight norm + beta=0.4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 | Y | Y | Y | 5035 |

It seems that weight scaling might be more consistent.

Trinkle23897 replied:
> I suspect that other implementations have done some weight scaling on the fly.

Agree, but I personally do not favor the scaling fix (1e9). Adding such a magic number and this cannot find in all other implementations, that's pretty weird... Maybe we miss something?

I dig into stable-baselines implementation once again and they indeed have the eps-hack, but instead on DQN policy side: 
https://github.com/hill-a/stable-baselines/blob/f9be700e26a7eca9ec25e04003479910100d6662/stable_baselines/deepq/dqn.py#L299

nuance1979 replied:
> > I suspect that other implementations have done some weight scaling on the fly.
> 
> Agree, but I personally do not favor the scaling fix (1e9). Adding such a magic number and this cannot find in all other implementations, that's pretty weird... Maybe we miss something?
> 
I agree scaling by a magic number is bad. I would rather do weight normalization.

> I dig into stable-baselines implementation once again and they indeed have the eps-hack, but instead on DQN policy side:
> https://github.com/hill-a/stable-baselines/blob/f9be700e26a7eca9ec25e04003479910100d6662/stable_baselines/deepq/dqn.py#L299
>
Great! So we'll definitely keep the eps hack. Having it on the buffer side is obviously better.

I tried `--training-num 4` and got some interesting results:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| sanity check       | N | N | N | 16530 (C51 baseline: 16245) |
| prio only              | Y | N | N | 3982.5 |
| w/ weight scaling + beta=0.4 | Y | N | N | 16582.5 |
| weight scaling + beta=0.4 | Y | Y | Y |  11625  |
| w/ weight norm + beta=0.4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 | Y | Y | Y | 5035 |
| w/ weight norm + beta=0.4 + training-num=4 | Y | N | N | 17562.5 |
| weight norm + beta=0.4 + training-num=4 | Y | Y | Y | 14925 |

I'm a little surprised by the last result, which seems to suggest that weight normalization with proper parameters does work. Since we don't like scaling by a magic number, I think we should keep the weight normalization trick.

Trinkle23897 replied:
> Here the weights returned are normalized by max:
> https://github.com/Kaixhin/Rainbow/blob/master/memory.py#L154

That's not returned weight, that's sampled weight

nuance1979 replied:
I think it does. See https://github.com/Kaixhin/Rainbow/blob/9ff5567ad1234ae0ed30d8471e8f13ae07119395/agent.py#L63
```python
idxs, states, actions, returns, next_states, nonterminals, weights = mem.sample(self.batch_size)
```

Trinkle23897 commented:
Why did you move weight norm to policy side?
![2021-07-19 09-17-30屏幕截图](https://user-images.githubusercontent.com/8189182/126089198-b3e3f3d3-7b7c-45ef-8f53-6fcc79a73e99.png)

nuance1979 commented:
> Why did you move weight norm to policy side?
>
I thought for backward compatibility it might be better to put the weight norm hack on the policy side. Basically it's up to the policy to decide whether it wants to use the weight norm hack or not.

I don't have a strong opinion about it so if you insist, I can move it back to the buffer side.

Trinkle23897 commented:
So how about adding an extra argument in prio-buffer and set it default to False?

Trinkle23897 commented:
I have some new findings

1. https://github.com/deepmind/dqn_zoo/blob/3bd1bbf7362e3310611903752b0dd4032646d462/dqn_zoo/prioritized/run_atari.py#L84 The weight normalization is enabled by default with author's setting.
2. about the weight processing: https://github.com/deepmind/acme/blob/9f82160644e6ef34c42c30a93e93651277f8b8fc/acme/agents/jax/dqn/losses.py#L66-L69

nuance1979 commented:
> I have some new findings
> 
> 1. https://github.com/deepmind/dqn_zoo/blob/3bd1bbf7362e3310611903752b0dd4032646d462/dqn_zoo/prioritized/run_atari.py#L84 The weight normalization is enabled by default with author's setting.
> 2. about the weight processing: https://github.com/deepmind/acme/blob/9f82160644e6ef34c42c30a93e93651277f8b8fc/acme/agents/jax/dqn/losses.py#L66-L69
> 
> ps: https://github.com/deepmind/dqn_zoo/ can be treated as author's implementation (based on Lua originally).

Great! I have changed the default value of weight_norm option to True.

I read the code in https://github.com/deepmind/dqn_zoo/ carefully and felt that the only remaining difference was beta annealing. So I tried it on Seaquest, where fixed beta resulted in bad performance. The results:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| C51 baseline       | N | N | N | 6226 |
| weight norm + beta=0.4              | Y | Y | Y | 2356 |
| weight norm + beta from 0.4 to 1 | Y | Y | Y | 10576 |

It seems annealing beta is essential for Rainbow. I added a `set_beta()` method to prio buffer to enable beta annealing. Do you think that's ok?

```python
def set_beta(self, beta: float) -> None:
    self._beta = beta
```

Currently the annealing schedule is quite arbitrary. I can change it to mimic what dqn_zoo does.

Trinkle23897 commented:
Yes that's totally okay

szrlee commented:
Will this commit be merged in the future?

Trinkle23897 commented:
Yes

Trinkle23897 commented:
@nuance1979 is it ready now...?

nuance1979 commented:
> @nuance1979 is it ready now...?

Sorry for the delay. 

After fixing the weight normalization and beta annealing, some tasks (Enduro, SpaceInvaders, etc.) still get terrible results so I tried hard to figure out why. Let me summarize my current findings:

I compared the current NoisyLinear implementation with https://github.com/deepmind/dqn_zoo/ and found two differences:

1. A different but equivalent way of adding the noise;
2. dqn_zoo removes the bias term for the second NoisyLinear layer.

I tried to align with dqn_zoo implementations regarding the above two but found no meaningful differences in Enduro performance.

So I tried turning off some features of Rainbow: 

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| Enduro C51 baseline (copied from README.md)  | N | N | N | 1032 |
| weight norm + beta from 0.4 to 1 | Y | Y | Y | 450.2 |
| - prio buffer | N | Y | Y | 487.2 |
| - prio buffer - dueling | N | Y | N | 354.9 |
| - prio buffer - dueling - noisy | N | N | N | 1459.7 |
| - prio buffer - noisy | N | N | Y | 1469 |
| - noisy | Y | N | Y | 1645.6 |

Therefore for Enduro, NoisyLinear layer hurts the performance. I suspect the same would be true for other low-performing tasks.

I feel that I have exhausted ideas to explore further. I could add "--no-noisy" as a task-specific parameter for Enduro (and potentially more tasks with similar behavior) or I could keep the low-performing numbers for the sake of consistency.

What do you think? @Trinkle23897

Trinkle23897 commented:
That's fine for --no-noisy. But have you ever tried --training-num=4 ? I suspect it suffers from policy lag.

nuance1979 commented:
> That's fine for --no-noisy. But have you ever tried --training-num=4 ? I suspect it suffers from policy lag.

I tried it before and it didn't make a difference. But things have changed a lot since then. I'll kick off an experiment to confirm.

nuance1979 commented:
> > That's fine for --no-noisy. But have you ever tried --training-num=4 ? I suspect it suffers from policy lag.
> 
> I tried it before and it didn't make a difference. But things have changed a lot since then. I'll kick off an experiment to confirm.

I got the result:

| Experiment         | Priority Buffer | NoisyLinear  | Dueling | Best reward |
| :---------------- | --------------:| -------------:| ------:| ------:|
| Enduro C51 baseline (copied from README.md)  | N | N | N | 1032 |
| weight norm + beta from 0.4 to 1 | Y | Y | Y | 450.2 |
| + training_num_4 | Y | Y | Y | 957.3 |
| - noisy (training_num_10) | Y | N | Y | 1645.6 |

Better than default ("--training-num=10") but still far worse than "--no-noisy".

Trinkle23897 commented:
okay that's fine. feel free to add `--no-noisy`

nuance1979 commented:
I might have found a bug in my code which could cause NoisyLinear not working correctly. I'm running experiments to confirm.

Trinkle23897 commented:
I have another unrelated issue: usually the atari network's feature part is end up by linear(3136, 512) instead of nn.flatten. I see only https://github.com/ku2482/fqf-iqn-qrdqn.pytorch use the latter setting, is that correct?

nuance1979 commented:
> I have another unrelated issue: usually the atari network's feature part is end up by linear(3136, 512) instead of nn.flatten. I see only https://github.com/ku2482/fqf-iqn-qrdqn.pytorch use the latter setting, is that correct?

I think that's correct. Otherwise it will not match the input shape of the following linear layer. I found the following line, which is just another way of flattening, in the repo where there is no `nn.Flatten` in the model itself:

https://github.com/Kaixhin/Rainbow/blob/9ff5567ad1234ae0ed30d8471e8f13ae07119395/model.py#L71

```
    x = x.view(-1, self.conv_output_size)
```

Trinkle23897 commented:
well I mean in on-policy atari setting they use the following:

```python
ActorCriticCnnPolicy(
  (features_extractor): NatureCNN(
    (cnn): Sequential(
      (0): Conv2d(4, 32, kernel_size=(8, 8), stride=(4, 4))
      (1): ReLU()
      (2): Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2))
      (3): ReLU()
      (4): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1))
      (5): ReLU()
      (6): Flatten(start_dim=1, end_dim=-1)
    )
    (linear): Sequential(
      (0): Linear(in_features=3136, out_features=512, bias=True)
      (1): ReLU()
    )
  )
  (action_net): Linear(in_features=512, out_features=6, bias=True)
  (value_net): Linear(in_features=512, out_features=1, bias=True)
)
```

nuance1979 commented:
> well I mean in on-policy atari setting they use the following:
> 
> ```python
> ActorCriticCnnPolicy(
>   (features_extractor): NatureCNN(
>     (cnn): Sequential(
>       (0): Conv2d(4, 32, kernel_size=(8, 8), stride=(4, 4))
>       (1): ReLU()
>       (2): Conv2d(32, 64, kernel_size=(4, 4), stride=(2, 2))
>       (3): ReLU()
>       (4): Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1))
>       (5): ReLU()
>       (6): Flatten(start_dim=1, end_dim=-1)
>     )
>     (linear): Sequential(
>       (0): Linear(in_features=3136, out_features=512, bias=True)
>       (1): ReLU()
>     )
>   )
>   (action_net): Linear(in_features=512, out_features=6, bias=True)
>   (value_net): Linear(in_features=512, out_features=1, bias=True)
> )
> ```
> 
> Could you please double-check the author's implementation on a series of offline-rl atari settings you previously implemented, together with fqf and iqn?

I see. The above structure essentially shared one more linear layer of (3136, 512) between action and value nets. I checked a few repos:

- dqn_zoo does NOT share this layer. [link1](https://github.com/deepmind/dqn_zoo/blob/master/dqn_zoo/networks.py#L235)  and [link2](https://github.com/deepmind/dqn_zoo/blob/master/dqn_zoo/networks.py#L243)
- Dopamine does. [link1](https://github.com/google/dopamine/blob/master/dopamine/jax/networks.py#L390) and [link2](https://github.com/google/dopamine/blob/master/dopamine/jax/networks.py#L391)
- Kaixhin's rainbow does NOT share this layer. [link1](https://github.com/Kaixhin/Rainbow/blob/master/model.py#L64) and [link2](https://github.com/Kaixhin/Rainbow/blob/master/model.py#L65)
- fqf-iqn-qrdqn's rainbow does NOT share this layer. [link1](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/masterr/fqf_iqn_qrdqn/network.py#L143) and [link2](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/master/fqf_iqn_qrdqn/network.py#L148)
- My current Rainbow implementation does NOT share this layer.

So for Rainbow's action and value nets, we are with the majority.

For IQN, the equivalent question is whether the CosNet has input size 3136 or 512 (i.e., after one linear layer of (3136, 512)). All the repos above has input size 3136, following the original paper. Same for FQF.

For offline-rl methods, CQL doesn't have this problem since no two heads share layers; CRR does have this issue since there are actor and critic nets. However, I can't find the reference implementation from the authors.

nuance1979 commented:
> I might have found a bug in my code which could cause NoisyLinear not working correctly. I'm running experiments to confirm.

I have fixed a bug due to a misunderstanding of the role NoisyLinear plays in Rainbow model. Basically I disabled explore_noise when NoisyLinear layers were used. Therefore the bad results with NoisyLinear layer were mainly due to the absence of the exploration. After the fix, the results with NoisyLinear are comparable to, but not always better than, the results without. Since we only have a single run, I wouldn't draw any conclusion based on it. Now I will not use "--no-noisy" option for the reported results in README.md .

There is one exception: Seaquest. When I accidentally disabled explore_noise, I got much higher results (~16000 vs ~2300). I couldn't figure out the reason.

Anyway I think this PR is ready to merge.

Trinkle23897 commented:
Cool, I'll take a look this weekend

wptoux commented:
There are some differences in the implementation of the function f(x). In this repo, the ```x = torch.randn(x.size(0), device=x.device)``` will override the input parameter x.

https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/11d70bb428e449fe5384654c05e4ab2c3bbdd4cd/fqf_iqn_qrdqn/network.py#L218-219

https://github.com/thu-ml/tianshou/blob/3592f45446e6cc98423df2f1c28d8ca0ef2be821/tianshou/utils/net/discrete.py#L349-L351

Trinkle23897 commented:
> There are some differences in the implementation of the function f(x). In this repo, the x = torch.randn(x.size(0), device=x.device) will override the input parameter x.

> https://github.com/ku2482/fqf-iqn-qrdqn.pytorch/blob/11d70bb428e449fe5384654c05e4ab2c3bbdd4cd/fqf_iqn_qrdqn/network.py#L218-219

> https://github.com/thu-ml/tianshou/blob/3592f45446e6cc98423df2f1c28d8ca0ef2be821/tianshou/utils/net/discrete.py#L349-L351

@nuance1979 is that the cause of bad performance? Could you please check it when you're free? Many thanks!
